File: src/sim/chem-ids.ts
// Chemical identity layer: the molecular pool shape, the locked
// chem-slot id space, and the slot/role count constants. Pure
// declarations -- no engine state, no rng, no table build -- so the
// reaction engine, regional field, creatures and serializer can all
// import ids/types without a value cycle back through sim.ts. The
// CHEMICALS table + LUTs (which depend on spawn ranking) stay in
// sim.ts for now and import from here.

// Per-cell molecular pool. ATP itself lives on the Creature as
// `energy`; every other named species in the chemistry lives here.
// All quantities are in the same mass units as reserves, so reactions
// are mass-conserving and cell volume is total mass.
export interface Molecules {
  adp: number;          // ATP's discharged form
  glucose: number;
  fattyAcid: number;
  aminoAcid: number;
  chlorophyll: number;
  enzyme: number;
  o2: number;
  co2: number;
  minerals: number;
  waste: number;
  mrna: number;
  biopolymer: number;
  membrane: number;
  // Photoreceptor band variants (visible / long-penetrating / depth-
  // invariant surface). Cells choose bands to invest in via the
  // unified SYNTH op; each band has its own activated chem.
  photoreceptorVisible: number;
  photoreceptorLong: number;
  photoreceptorSurface: number;
  activatedPhotoVisible: number;
  activatedPhotoLong: number;
  activatedPhotoSurface: number;
  // Retired CHEMO-branch receptor chems (the chemo gradient sense was
  // superseded by universal SENSE_OUT). Their ids/columns are being
  // REPURPOSED into the new sensory modalities (SENSES_PLAN.md) rather
  // than left inert: chemoreceptorFa(21)->phreceptor (pH/acidity sense).
  // The remaining biopolymer/minerals/marker0 slots are repurposed in
  // later modality commits (electric/vibration/light).
  electroreceptor: number;
  vibroreceptor: number;
  phreceptor: number; // ex-chemoreceptorFa (id 21): acidity receptor
  chemoreceptorMarker0: number;
  activatedElectroX: number;
  activatedElectroY: number;
  activatedVibX: number;
  activatedVibY: number;
  activatedPh: number; // ex-activatedChemoFaX (id 27): acidity signal (scalar)
  activatedChemoFaY: number; // dead spare (id 28), future repurpose

  activatedLightX: number;
  activatedLightY: number;
  // Mechano / thermo / magneto.
  mechanoreceptor: number;
  activatedMechX: number;
  activatedMechY: number;
  thermoreceptor: number;
  activatedThermo: number;
  magnetoreceptor: number;
  activatedMagX: number;
  activatedMagY: number;
  // Phase K ADHERE / REPAIR rework: chemistry-mediated bonding and
  // somatic-mutation suppression.
  bondChem: number;
  repairChem: number;
  // Identity markers (last so additions above don't shift their ids).
  marker0: number;
  marker1: number;
  marker2: number;
  marker3: number;
  // ATP. Path 1: ATP is now a first-class chemical (was the
  // per-creature `energy` scalar). It is mass-conserved like every
  // other molecule and participates in the chem machinery (transport
  // / reactions). The Creature.energy accessor + store.energy array
  // are aliased onto this column, so all existing energy code paths
  // are unchanged. Appended last (named id 45) so chems 0..44 keep
  // their ids and the genome ABI (%CHEMICAL_COUNT) is unchanged.
  atp: number;
}

export const MOLECULE_IDS: ReadonlyArray<keyof Molecules> = [
  "adp", "glucose", "fattyAcid", "aminoAcid", "chlorophyll", "enzyme",
  "o2", "co2", "minerals", "waste", "mrna",
  "biopolymer", "membrane",
  "photoreceptorVisible", "photoreceptorLong", "photoreceptorSurface",
  "activatedPhotoVisible", "activatedPhotoLong", "activatedPhotoSurface",
  "electroreceptor", "vibroreceptor", "phreceptor", "chemoreceptorMarker0",
  "activatedElectroX", "activatedElectroY",
  "activatedVibX", "activatedVibY",
  "activatedPh", "activatedChemoFaY",
  "activatedLightX", "activatedLightY",
  "mechanoreceptor", "activatedMechX", "activatedMechY",
  "thermoreceptor", "activatedThermo",
  "magnetoreceptor", "activatedMagX", "activatedMagY",
  "bondChem", "repairChem",
  "marker0", "marker1", "marker2", "marker3",
  "atp",
];

export function emptyMolecules(): Molecules {
  // Build from MOLECULE_IDS so adding a new entry above auto-zero-
  // initializes here. Cheap (one allocation per call); used mostly
  // by snapshots / death release / test scaffolds.
  const m = {} as Molecules;
  for (const k of MOLECULE_IDS) (m as unknown as Record<string, number>)[k] = 0;
  return m;
}

// String key -> index in MOLECULE_IDS array. Maps between named chem
// ids and Molecules field positions.
export const MOLECULE_INDEX: Record<keyof Molecules, number> =
  {} as Record<keyof Molecules, number>;
for (let i = 0; i < MOLECULE_IDS.length; i++) MOLECULE_INDEX[MOLECULE_IDS[i]] = i;

export const CHEMICAL_COUNT = 96;
export const NAMED_CHEMICAL_COUNT = 46;
// Order maps to chemCols[0..NAMED_CHEMICAL_COUNT-1]. The CHEM_* slot
// constants below MUST match this order; see CHEMISTRY_OVERHAUL.md
// phase K for the locked layout.
export const NAMED_CHEMICALS: ReadonlyArray<keyof Molecules> = [
  "o2", "co2", "glucose", "aminoAcid", "fattyAcid", "minerals", "adp",
  "waste", "chlorophyll", "enzyme", "mrna",
  "biopolymer", "membrane",
  "photoreceptorVisible", "photoreceptorLong", "photoreceptorSurface",
  "activatedPhotoVisible", "activatedPhotoLong", "activatedPhotoSurface",
  "electroreceptor", "vibroreceptor", "phreceptor", "chemoreceptorMarker0",
  "activatedElectroX", "activatedElectroY",
  "activatedVibX", "activatedVibY",
  "activatedPh", "activatedChemoFaY",
  "activatedLightX", "activatedLightY",
  "mechanoreceptor", "activatedMechX", "activatedMechY",
  "thermoreceptor", "activatedThermo",
  "magnetoreceptor", "activatedMagX", "activatedMagY",
  "bondChem", "repairChem",
  "marker0", "marker1", "marker2", "marker3",
  "atp",
];

// Slot indices for special handling (engine-managed ATP/ADP, etc.).
export const CHEM_O2 = 0;
export const CHEM_CO2 = 1;
export const CHEM_GLU = 2;
export const CHEM_AA = 3;
export const CHEM_FA = 4;
export const CHEM_MIN = 5;
export const CHEM_ADP = 6;
export const CHEM_WASTE = 7;
// chlorophyll, enzyme, mrna have specific roles as rate multipliers:
//   chl   -> photosynth (mandatory: no chl -> no photosynth)
//   ribo  -> all biosynth reactions (mandatory: no ribo -> no biosynth)
//   enz   -> catabolize (mandatory: no enz -> no digestion of biopolymer)
export const CHEM_CHL = 8;
export const CHEM_ENZ = 9;
export const CHEM_MRNA = 10;
export const CHEM_BIOPOLYMER = 11;
export const CHEM_MEMBRANE = 12;
// Phase K-1 layout: receptors split into band / per-target variants,
// each with paired activated chems. Order matches NAMED_CHEMICALS.
export const CHEM_PHOTORECEPTOR_VISIBLE = 13;
export const CHEM_PHOTORECEPTOR_LONG = 14;
export const CHEM_PHOTORECEPTOR_SURFACE = 15;
export const CHEM_ACT_PHOTO_VISIBLE = 16;
export const CHEM_ACT_PHOTO_LONG = 17;
export const CHEM_ACT_PHOTO_SURFACE = 18;
export const CHEM_ELECTRORECEPTOR = 19;
export const CHEM_VIBRORECEPTOR = 20;
export const CHEM_PHRECEPTOR = 21; // ex-CHEM_CHEMORECEPTOR_FA: acidity receptor
export const CHEM_CHEMORECEPTOR_MARKER0 = 22;
export const CHEM_ACT_ELECTRO_X = 23;
export const CHEM_ACT_ELECTRO_Y = 24;
export const CHEM_ACT_VIB_X = 25;
export const CHEM_ACT_VIB_Y = 26;
export const CHEM_ACT_PH = 27; // ex-CHEM_ACT_CHEMO_FA_X: acidity signal (scalar)
export const CHEM_ACT_CHEMO_FA_Y = 28;
export const CHEM_ACT_LIGHT_X = 29;
export const CHEM_ACT_LIGHT_Y = 30;
export const CHEM_MECHANORECEPTOR = 31;
export const CHEM_ACT_MECH_X = 32;
export const CHEM_ACT_MECH_Y = 33;
export const CHEM_THERMORECEPTOR = 34;
export const CHEM_ACT_THERMO = 35;
export const CHEM_MAGNETORECEPTOR = 36;
export const CHEM_ACT_MAG_X = 37;
export const CHEM_ACT_MAG_Y = 38;
export const CHEM_BOND = 39;
export const CHEM_REPAIR = 40;
export const CHEM_MARKER0 = 41;
// Markers occupy 41..44; marker0 has a constant since the
// chemoreceptor system targets it specifically.
// ATP: named id 45 (appended after markers; first ex-generic slot).
// CHEMICAL_COUNT stays 96 so the genome ABI is unchanged.
export const CHEM_ATP = 45;
export const MRNA_REF = 5;
// Lowered 5 -> 2: photosynthesis + photophosphorylation scale by
// min(1, chl/CHL_REF), and a fresh founder seeds only chl=1, so at
// CHL_REF=5 it ran light-harvesting at 20% -- too little ATP to afford
// building more chlorophyll (synth_chl costs -8 ATP), a bootstrap
// deadlock that left <0.1% of cells able to ignite autotrophy. At 2 the
// seed chl=1 drives 50% light-harvesting (~8x more cells ignite), while
// reaching full rate still requires investing in chl (SYNTH_CHL), so
// autotrophy stays an evolved trait -- just a reachable one. (Note: this
// broadens ignition but does NOT by itself make the world self-
// sustaining; seedless descendants still rarely reach the reproduce
// threshold -- see the founder-subsidy finding.)
export const CHL_REF = 2;
export const ENZ_REF = 5;
export const GENERIC_CHEMICAL_COUNT = CHEMICAL_COUNT - NAMED_CHEMICAL_COUNT;

// CHEM_NAMED_MOL_IDX[k] = molCols index of the named chemical at
// chemCols[k]. Resolved against MOLECULE_INDEX (populated above).
export const CHEM_NAMED_MOL_IDX: number[] = NAMED_CHEMICALS.map((n) => MOLECULE_INDEX[n]);

File: server/crates/evosim-engine/src/chem_ids.rs
//! Chemical identity layer ported from `src/sim/chem-ids.ts`. Pure
//! declarations -- the molecular-pool shape, the locked chem-slot id
//! space, and the slot/role count constants. No engine state, no rng,
//! no table build. Every other engine module that names a chemical
//! imports these constants so the id space stays single-sourced.
//!
//! Order must match `NAMED_CHEMICALS` exactly; reordering would break
//! the bond-projection LUT and the genome ABI's `%CHEMICAL_COUNT`
//! arithmetic.

/// Total chemical slots (named + procedurally generated generics).
pub const CHEMICAL_COUNT: usize = 96;

/// Named slots at the head of the chem table. Generics live at
/// `NAMED_CHEMICAL_COUNT..CHEMICAL_COUNT`.
pub const NAMED_CHEMICAL_COUNT: usize = 46;

pub const GENERIC_CHEMICAL_COUNT: usize = CHEMICAL_COUNT - NAMED_CHEMICAL_COUNT;

// Named-chem slot constants. Order matches `NAMED_CHEMICALS` (and the
// TS file's CHEM_* exports). Don't renumber.

pub const CHEM_O2: usize = 0;
pub const CHEM_CO2: usize = 1;
pub const CHEM_GLU: usize = 2;
pub const CHEM_AA: usize = 3;
pub const CHEM_FA: usize = 4;
pub const CHEM_MIN: usize = 5;
pub const CHEM_ADP: usize = 6;
pub const CHEM_WASTE: usize = 7;
pub const CHEM_CHL: usize = 8;
pub const CHEM_ENZ: usize = 9;
pub const CHEM_MRNA: usize = 10;
pub const CHEM_BIOPOLYMER: usize = 11;
pub const CHEM_MEMBRANE: usize = 12;
pub const CHEM_PHOTORECEPTOR_VISIBLE: usize = 13;
pub const CHEM_PHOTORECEPTOR_LONG: usize = 14;
pub const CHEM_PHOTORECEPTOR_SURFACE: usize = 15;
pub const CHEM_ACT_PHOTO_VISIBLE: usize = 16;
pub const CHEM_ACT_PHOTO_LONG: usize = 17;
pub const CHEM_ACT_PHOTO_SURFACE: usize = 18;
pub const CHEM_ELECTRORECEPTOR: usize = 19;
pub const CHEM_VIBRORECEPTOR: usize = 20;
pub const CHEM_PHRECEPTOR: usize = 21;
pub const CHEM_CHEMORECEPTOR_MARKER0: usize = 22;
pub const CHEM_ACT_ELECTRO_X: usize = 23;
pub const CHEM_ACT_ELECTRO_Y: usize = 24;
pub const CHEM_ACT_VIB_X: usize = 25;
pub const CHEM_ACT_VIB_Y: usize = 26;
pub const CHEM_ACT_PH: usize = 27;
pub const CHEM_ACT_CHEMO_FA_Y: usize = 28;
pub const CHEM_ACT_LIGHT_X: usize = 29;
pub const CHEM_ACT_LIGHT_Y: usize = 30;
pub const CHEM_MECHANORECEPTOR: usize = 31;
pub const CHEM_ACT_MECH_X: usize = 32;
pub const CHEM_ACT_MECH_Y: usize = 33;
pub const CHEM_THERMORECEPTOR: usize = 34;
pub const CHEM_ACT_THERMO: usize = 35;
pub const CHEM_MAGNETORECEPTOR: usize = 36;
pub const CHEM_ACT_MAG_X: usize = 37;
pub const CHEM_ACT_MAG_Y: usize = 38;
pub const CHEM_BOND: usize = 39;
pub const CHEM_REPAIR: usize = 40;
pub const CHEM_MARKER0: usize = 41;
// Markers occupy 41..44; only marker0 has a constant since the
// chemoreceptor system targets it specifically.
pub const CHEM_ATP: usize = 45;

/// Mandatory-multiplier reference levels for the three role chems.
/// See `src/sim/chem-ids.ts` for the long-form comment on why CHL_REF
/// was lowered from 5 to 2.
pub const MRNA_REF: f32 = 5.0;
pub const CHL_REF: f32 = 2.0;
pub const ENZ_REF: f32 = 5.0;

/// Names indexed by chem slot. Single source of truth for the chem
/// label; nothing else in the engine should hard-code "o2" etc.
/// Mirrors the TS `NAMED_CHEMICALS` list but as a `match` so the
/// compiler refuses to silently let a slot drift.
pub fn slot_name(slot: usize) -> &'static str {
    match slot {
        CHEM_O2 => "o2",
        CHEM_CO2 => "co2",
        CHEM_GLU => "glucose",
        CHEM_AA => "aminoAcid",
        CHEM_FA => "fattyAcid",
        CHEM_MIN => "minerals",
        CHEM_ADP => "adp",
        CHEM_WASTE => "waste",
        CHEM_CHL => "chlorophyll",
        CHEM_ENZ => "enzyme",
        CHEM_MRNA => "mrna",
        CHEM_BIOPOLYMER => "biopolymer",
        CHEM_MEMBRANE => "membrane",
        CHEM_PHOTORECEPTOR_VISIBLE => "photoreceptorVisible",
        CHEM_PHOTORECEPTOR_LONG => "photoreceptorLong",
        CHEM_PHOTORECEPTOR_SURFACE => "photoreceptorSurface",
        CHEM_ACT_PHOTO_VISIBLE => "activatedPhotoVisible",
        CHEM_ACT_PHOTO_LONG => "activatedPhotoLong",
        CHEM_ACT_PHOTO_SURFACE => "activatedPhotoSurface",
        CHEM_ELECTRORECEPTOR => "electroreceptor",
        CHEM_VIBRORECEPTOR => "vibroreceptor",
        CHEM_PHRECEPTOR => "phreceptor",
        CHEM_CHEMORECEPTOR_MARKER0 => "chemoreceptorMarker0",
        CHEM_ACT_ELECTRO_X => "activatedElectroX",
        CHEM_ACT_ELECTRO_Y => "activatedElectroY",
        CHEM_ACT_VIB_X => "activatedVibX",
        CHEM_ACT_VIB_Y => "activatedVibY",
        CHEM_ACT_PH => "activatedPh",
        CHEM_ACT_CHEMO_FA_Y => "activatedChemoFaY",
        CHEM_ACT_LIGHT_X => "activatedLightX",
        CHEM_ACT_LIGHT_Y => "activatedLightY",
        CHEM_MECHANORECEPTOR => "mechanoreceptor",
        CHEM_ACT_MECH_X => "activatedMechX",
        CHEM_ACT_MECH_Y => "activatedMechY",
        CHEM_THERMORECEPTOR => "thermoreceptor",
        CHEM_ACT_THERMO => "activatedThermo",
        CHEM_MAGNETORECEPTOR => "magnetoreceptor",
        CHEM_ACT_MAG_X => "activatedMagX",
        CHEM_ACT_MAG_Y => "activatedMagY",
        CHEM_BOND => "bondChem",
        CHEM_REPAIR => "repairChem",
        CHEM_MARKER0 => "marker0",
        42 => "marker1",
        43 => "marker2",
        44 => "marker3",
        CHEM_ATP => "atp",
        _ => "generic", // 46..96 share a label; the chem table carries the real one
    }
}

/// True iff the chem at this slot is a *signal* chem (signed scalar
/// activation, not physical mass). Sensor activation chems are stored
/// in the same molecule pool but excluded from mass conservation.
pub fn is_signal(slot: usize) -> bool {
    matches!(
        slot,
        CHEM_ACT_PHOTO_VISIBLE
            | CHEM_ACT_PHOTO_LONG
            | CHEM_ACT_PHOTO_SURFACE
            | CHEM_ACT_ELECTRO_X
            | CHEM_ACT_ELECTRO_Y
            | CHEM_ACT_VIB_X
            | CHEM_ACT_VIB_Y
            | CHEM_ACT_PH
            | CHEM_ACT_CHEMO_FA_Y
            | CHEM_ACT_LIGHT_X
            | CHEM_ACT_LIGHT_Y
            | CHEM_ACT_MECH_X
            | CHEM_ACT_MECH_Y
            | CHEM_ACT_THERMO
            | CHEM_ACT_MAG_X
            | CHEM_ACT_MAG_Y
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn slot_names_align_with_constants() {
        assert_eq!(slot_name(CHEM_O2), "o2");
        assert_eq!(slot_name(CHEM_ATP), "atp");
        assert_eq!(slot_name(CHEM_PHRECEPTOR), "phreceptor");
        assert_eq!(slot_name(CHEM_ACT_MAG_X), "activatedMagX");
    }

    #[test]
    fn slot_counts_match_ts() {
        assert_eq!(CHEMICAL_COUNT, 96);
        assert_eq!(NAMED_CHEMICAL_COUNT, 46);
        assert_eq!(GENERIC_CHEMICAL_COUNT, 50);
    }

    #[test]
    fn signal_chems_excluded_from_mass() {
        // Every CHEM_ACT_* should report as signal; nothing else
        // should.
        assert!(is_signal(CHEM_ACT_LIGHT_X));
        assert!(!is_signal(CHEM_GLU));
        assert!(!is_signal(CHEM_ATP));
    }
}
